src/Microsoft.ML.Core/Microsoft.ML.Core.csproj

```diff
@@ -10,9 +10,9 @@
 
   <ItemGroup>
     <ProjectReference Include="..\Microsoft.ML.DataView\Microsoft.ML.DataView.csproj" />
-    
     <PackageReference Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutableVersion)" />
     <PackageReference Include="System.Memory" Version="$(SystemMemoryVersion)" />
+    <PackageReference Include="System.Drawing.Common" Version="$(SystemDrawingCommonPackageVersion)" />
```

Comment:
>SystemDrawingCommonPackageVersion [](start = 65, length = 33)

I think this is main reason why ImageDataType is in separate assembly.
I personally would prefer to make dataview system extensible, some kind of register mechanism in dataview type so we can have types necessary for certain workflow be separated from main nuget.

Reply:
The parsing framework now depends on image type. How can it be in another assembly? It sounds a circular dependency, right? In addition, refactoring ML.NET to support customized type doesn't sound a small task, so I think we need shiproom's approval.

---
In reply to: [273717323](https://github.com/dotnet/machinelearning/pull/3263#discussion_r273717323) [](ancestors = 273717323)

Reply:
>The parsing framework now depends on image type. 
What? [https://media1.tenor.com/images/d1fbf17571ea3260dee054ceebf65fd7/tenor.gif?itemid=8605657](https://media1.tenor.com/images/d1fbf17571ea3260dee054ceebf65fd7/tenor.gif?itemid=8605657)

All I'm trying to say is I'm not sure it's worth to move image type inside basic nugets for the price of adding another dependency.

I agree support for customized type is quite big task, and shouldn't be part of 1.0.
But it can be part of 1.1


---
In reply to: [273720375](https://github.com/dotnet/machinelearning/pull/3263#discussion_r273720375) [](ancestors = 273720375,273717323)

Reply:
This is a standard library downloaded by million times and also support .NET Core, so maybe adding it is fine?

---
In reply to: [273721913](https://github.com/dotnet/machinelearning/pull/3263#discussion_r273721913) [](ancestors = 273721913,273720375,273717323)

Reply:
Note: the kind of change you are making here is not possible to do after `1.0` ships. You cannot change namespaces on types - it is an API breaking change.

Also, if you move one type from one assembly to another, it is a breaking change (unless you add a `TypeForwardedTo` attribute).

Reply:
So, thanks so much for bringing this problem up. I definitely agree it's a real problem. That said, let me just give a hard "no" on this particular solution, not because it doesn't work for *this* custom type, but because there's no conceivable way it could work for the hypothetical next custom type (for the reasons @eerhardt has already mentioned).

Let me just propose the plausible workable alternative, which has the benefit of actually doing something we've successfully relied on for years and years. Our model loading logic is built on a system where the model loading/saving core relies in a fairly central nuget, but it does not have direct references. It does this through the component catalog. We also use this, for example, for custom mappings, where user code (which we obviously can't depend on!!) can register their own plugins. If we delve into the internal Microsoft usage of this tool, we also have the experience of our command line -- plenty of `ICommand` implementations rely upon the existence of things like trainers and transforms that they cannot possibly directly reference. I could imagine this same strategy -- the schema deinfition and prediction engine code having access to components that handle this "mapping" -- working quite similarly.

More specifically, I could easily imagine that when someone defines a new derived type from `StructuredDataViewType`, they have the *option* of registering a plugin to this schema definition object that our reflection-based mechanism relies on, to provide whatever the appropriate mapping logic is, to facilitate their reading-in and writing-out from ML pipelines. If they do, great. If not, so be it, we don't offer it.

If you don't like that `ComponentCatalog` based solution, you could even imagine in some hypothetical alternate mechanism for creating these prediction engine that takes the mapping logic *directly* as another (again optional) parameter... or works through an attribute on the property or field on the user class (with an abstract base class in the core assemblies and derived classes in the more leaf assemblies) to describe the logic. Or a static registration mechanism by which user code can tell this, "hey, when you see this StructuredDataViewType and a field/property with this .NET representation type, this is what you do."

The good news is that all this can be done post 1.0. The only reasonable solution I could imagine that could perhaps not be done post 1.0, is if we chose to make the logic of how to handle a type embedded in the `StructuredDataViewType` class itself -- that is, how to "safely" copy from one type to another, and such things as this. But even that could be done post-1.0, really. Imagine a new interface, `IAmATypeThatKnowsHowToReadAndWrite<T>`, that's something we could always add later I think, without it being a breaking change. Who knows.

Anyway, plenty of options I think, all of which can be done post 1.0 and are I *think* superior to this solution, if you'll forgive my saying so?

The happy part about that is, I don't think any of that needs to be done prior to 1.0. I could easily imagine, say, a new abstract class, or a new plugin class, to be -- just how, hypothetically, the custom mapper transform *could* have been implemented after 1.0 (it obviously wasn't, it was one of the first things we did, but, if we hypothetically hadn't it would not represent a breaking cahnge)